### PR TITLE
[SPARK-34977][SQL] LIST FILES/JARS/ARCHIVES cannot handle multiple arguments properly when at least one path is quoted

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes an issue that `LIST FILES/JARS/ARCHIVES path1 path2 ...` cannot list all paths if at least one path is quoted.
An example here.
```
ADD FILE /tmp/test1;
ADD FILE /tmp/test2;

LIST FILES /tmp/test1 /tmp/test2;
file:/tmp/test1
file:/tmp/test2

LIST FILES /tmp/test1 "/tmp/test2";
file:/tmp/test2
```

In this example, the second `LIST FILES` doesn't show `file:/tmp/test1`.

To resolve this issue, I modified the syntax rule to be able to handle this case.
I also changed `SparkSQLParser` to be able to handle paths which contains white spaces.

### Why are the changes needed?

This is a bug.
I also have a plan which extends `ADD FILE/JAR/ARCHIVE` to take multiple paths like Hive and the syntax rule change is necessary for that.

### Does this PR introduce _any_ user-facing change?

Yes. Users can pass quoted paths when using `ADD FILE/JAR/ARCHIVE`.

### How was this patch tested?

New test.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 1]
```diff
@@ -952,6 +952,72 @@ class HiveQuerySuite extends HiveComparisonTest with SQLTestUtils with BeforeAnd
...
+      Files.write(file4.toPath, "file7".getBytes)
+      Files.write(file5.toPath, "file8".getBytes)
+
+      val jarFile4 = new File(dir, "test4.jar")
+      val jarFile5 = new File(dir, "test5.jar")
+      TestUtils.createJar(Seq(file7), jarFile4)
+      TestUtils.createJar(Seq(file8), jarFile5)
+
+      sql(s"ADD JAR ${jarFile4.getAbsolutePath}")
+      sql(s"ADD JAR ${jarFile5.getAbsolutePath}")
```
Comment (sarutak) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:1012]: Unlike `ADD FILE "path"` and `ADD ARCHIVE "PATH"`, we cannot execute `ADD JAR "path"` when the path contains whitespaces.
I think it's a bug and #32052 will fix this issue.

**Comment by sarutak** [comment]
retest this please.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:1012]: Got it. @sarutak .

**Comment by dongjoon-hyun** [comment]
#32052 is merged now. Could you rebase this to the master, @sarutak ?

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
cc: @HyukjinKwon

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 2]
```diff
@@ -67,17 +65,16 @@ case class ListFilesCommand(files: Seq[String] = Seq.empty[String]) extends Runn
...
-      files.map { f =>
-        val uri = new URI(f)
-        val schemeCorrectedPath = uri.getScheme match {
-          case null | "local" => new File(f).getCanonicalFile.toURI.toString
-          case _ => f
-        }
-        new Path(schemeCorrectedPath).toUri.toString
-      }.collect {
-        case f if fileList.contains(f) => f
-      }.map(Row(_))
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala:80]: @sarutak, what's the diff here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 3]
```diff
@@ -67,17 +65,16 @@ case class ListFilesCommand(files: Seq[String] = Seq.empty[String]) extends Runn
...
+      if (files.size > 0) {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala:68]: seems like there's a mistake in indentation

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 4]
```diff
@@ -67,17 +65,16 @@ case class ListFilesCommand(files: Seq[String] = Seq.empty[String]) extends Runn
...
   override def run(sparkSession: SparkSession): Seq[Row] = {
     val fileList = sparkSession.sparkContext.listFiles()
-    if (files.size > 0) {
-      files.map { f =>
-        val uri = new URI(f)
-        val schemeCorrectedPath = uri.getScheme match {
-          case null | "local" => new File(f).getCanonicalFile.toURI.toString
-          case _ => f
-        }
-        new Path(schemeCorrectedPath).toUri.toString
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala:77]: @sarutak, I think `listFiles` is always URIs if I am not mistaken. If they are all canonical URIs, I believe we don't have to switch `Path` (?).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 5]
```diff
@@ -96,7 +93,7 @@ case class ListJarsCommand(jars: Seq[String] = Seq.empty[String]) extends Runnab
     val jarList = sparkSession.sparkContext.listJars()
     if (jars.nonEmpty) {
       for {
-        jarName <- jars.map(f => new Path(f).getName)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala:99]: So, I think this is fine too (?)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 4 above)
Comment (sarutak) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala:77]: Actually, we can get a wrong result with `new Path(schemeCorrectedPath).toUri.toString`.
The constructor of `Path` which takes a `String` parameter doesn't handle URI formed string properly.
For example, "file:///test%20file.txt" is already encoded and valid URI but the constructor handles such URI as **NOT** encoded.
As a result, `new Path("file:///path/to/test%20file.txt").toUri.toString` returns `file:///path/to/test%2020file.txt` though `file:///path/to/test%20file.txt` is expected.

**Review by HyukjinKwon — COMMENTED** [review]
Looks okay

**Comment by sarutak** [comment]
Thanks all. Merged into `master`.

**Pull request closed without merging**